README.md
# HDraCor -- Type Safe Functional Drama Analysis in Haskell #

`HDraCor` is a wrapper around the API of
[dracor.org](https://dracor.org) for the
[Haskell](https://www.haskell.org/) programming language. It offers
functions for parsing the JSON data returned by the API functions into
Haskell data types.

## Installation ##

[Stack](https://haskellstack.org/), the tool for
building Haskell projects, is required. After installing `stack` just
clone this repo, `cd` into it and do:

	stack build

This will take some time and download quite a number of libraries. You
can run the test suite by

	stack test


## Usage of the library ##

`HDraCor` is minimal. Therefore it does not define a specific HTTP
library, but abstracts the HTTP layer away in a way seen in many other
functional programming libraries. All the library does is parsing JSON
to the records defined in `Text.DraCor.Types`. You can use
e.g. ['http-conduit'](https://hackage.haskell.org/package/http-conduit-2.3.7.3/docs/Network-HTTP-Conduit.html)
for HTTP:

	import Network.HTTP.Conduit
	import qualified Data.ByteString.Lazy as B

	fetch :: String -> IO B.ByteString
	fetch path = simpleHttp $ "https://dracor.org/api" ++ path


Now you can pass that `fetch` function to the functions from
`Text.DraCor.API` that wrap the API. E.g.:

	import Text.DraCor.API

	corpus <- getCorpus fetch "ger"

The result is wrapped in a `Maybe` functor and is `Nothing` if the
body of the HTTP response could not be parsed.

See the [`crawler`](crawler/app/Main.hs) for an example usage of the
library.

## Tolerant and intolerant parsers ##

The library offers two sets of JSON parsers: One set is fault-tolerant
and will succeed parsing despite issues #83-89 of
[dracor-api](https://github.com/dracor-org/dracor-api). The other
one--which is still under construction--is intolerant an can be used
for application testing.

Caveat: The two sets live in the modules `Text.DraCor.TolerantJSON` and
`Text.DraCor.IntolerantJSON` respectively and are *orphan
instances*. So, don't try to import both of them!

## Crawler ##

There is a crawler in the [`crawler`](./crawler) folder. It is
intended for testing the API and as a demo application. It can be
built and installed with

	stack build
	stack install

Then do

	crawldracor -h

to see what it can do.

## License

Licensed under either of:

- [BSD-3-Clause license](https://opensource.org/licenses/BSD-3-Clause)
- [Apache License, version 2.0](https://opensource.org/licenses/Apache-2.0)

As a user, you may use this code under either license, at your option.


ChangeLog.md
# Changelog for hdracor

## Unreleased changes


test/Spec.hs
{-# OPTIONS_GHC -F -pgmF htfpp #-}
module Main where

import Test.Framework
import Test.Framework.BlackBoxTest

import {-@ HTF_TESTS @-} Test.Text.DraCor.Tolerant
--import {-@ HTF_TESTS @-} Test.Text.DraCor.Intolerant

main = htfMain htf_importedTests


test/Test/Text/DraCor/Types.hs
{-# OPTIONS_GHC -F -pgmF htfpp #-}
{-# LANGUAGE OverloadedStrings #-}
module Test.Text.DraCor.Types
  where

import Test.Framework
import Data.Maybe
import Data.Aeson
import qualified Data.ByteString.Lazy as B
import Control.Monad

import Text.DraCor.Types
import Text.DraCor.IntolerantJSON

-- * low level: logical parts of data returned by the api

test_scene = do
  s <- B.readFile "test/examples/scene1.json"
  assertEqual True (isJust (decode s :: Maybe Scene))
  assertEqual
    (Just ["dietrich", "helldrungen", "erich"])
    (fmap scnSpeakers (decode s :: Maybe Scene))
  
test_sceneNoSpeakers = do
  s <- B.readFile "test/examples/scene2.json"
  assertEqual True (isJust (decode s :: Maybe Scene))



-- * high level: data presented by the api

test_info = do
  s <- B.readFile "test/examples/info"
  assertEqual True (isJust (decode s :: Maybe Info))
  assertEqual (Just "DraCor API")  (fmap apiName $ (decode s :: Maybe Info))
  
test_corpora = do
  s <- B.readFile "test/examples/corpora"
  assertEqual True (isJust (decode s :: Maybe [Corpus]))

test_corporaGer = do
  s <- B.readFile "test/examples/corpora-ger"
  assertEqual True (isJust (decode s :: Maybe Corpus))

-- see issue #85 of dracor-api
test_corporaGerMetadata = do
  s <- B.readFile "test/examples/corpora-ger-metadata"
  assertEqual True (isJust (decode s :: Maybe [Play]))

test_corporaGerPlayAlbertiBrot = do
  s <- B.readFile "test/examples/corpora-ger-play-alberti-brot"
  assertEqual True (isJust (decode s :: Maybe Play))

-- see issue #87 of the dracor-api
test_corporaGerPlayAlbertiBrotMetrics = do
  s <- B.readFile "test/examples/corpora-ger-play-alberti-brot-metrics"
  assertEqual True (isJust (decode s :: Maybe Play))




crawler/app/Main.hs
module Main where

import Options.Applicative
import Network.HTTP.Conduit
import Conduit (runConduit)
import Control.Monad.Trans.Resource (runResourceT)
import qualified Data.ByteString.Lazy as B
import Control.Monad
import Data.Maybe
import qualified Data.Text as T
import Control.Concurrent

import Text.DraCor.Types
import Text.DraCor.API


data Opts = Opts
  { url :: String
  , entry :: Entry
  , verbosity :: Verbosity
  , delay :: Int
  }

data Entry
  = Corpora
  | SingleCorpus String
  | SinglePlay String String

data Verbosity
  = Normal
  | Verbose


opts_ :: Parser Opts
opts_ = Opts
  <$> argument str (metavar "URL"
                    <> help "Base URL of dracor API."
                    <> showDefault
                    <> value "https://dracor.org/api")
  <*> ((flag Corpora Corpora
        (long "corpora"
         <> short 'a'
         <> help "Crawl all corpora and everything in it. (Default)"))
       <|>
       (SingleCorpus
        <$> strOption (long "corpus"
                       <> short 'c'
                       <> help "Crawl a single corpus and everything in it."
                       <> metavar "CORPUS"))
       <|>
       (SinglePlay
        <$> strOption (long "play"
                       <> short 'p'
                       <> help "Crawl a single play."
                       <> metavar "CORPUS")
        <*> argument str (metavar "PLAY"
                          <> help "The play's name.")))
  <*> ((flag Normal Normal
        (long "normal"
         <> short 'n'
         <> help "Normal verbosity (Default)."))
       <|>
       (flag' Verbose
        (long "verbose"
         <> short 'v'
         <> help "Verbose output.")))
  <*> option auto (long "timeout"
                   <> short 't'
                   <> help "Timeout between http requests."
                   <> metavar "MICROSECONDS"
                   <> value 500000
                   <> showDefault)


main = execParser opts >>= run
  where opts = info (helper <*> opts_)
          (fullDesc
           <> progDesc
           "crawldracor crawls a dracor-api for application testing."
           <> header "crawldracor  - crawl the api of a dracor endpoint.")


fetch' :: String -> String -> IO B.ByteString
-- fetch url path = do
--   request <- parseRequest $ url ++ path
--   manager <- newManager tlsManagerSettings
--   return $ responseBody $ httpLbs request manager
fetch' url path = simpleHttp $ url ++ path

fetch :: Opts -> (String -> IO B.ByteString)
fetch opts = fetch' $ url opts

run :: Opts -> IO ()
run opts@(Opts _ Corpora _ _) = crawlCorpora opts
run opts@(Opts _ (SingleCorpus name) _ _) = crawlCorpus opts name
run opts@(Opts _ (SinglePlay corpusName playName) _ _) = crawlPlay opts corpusName playName

crawlCorpora :: Opts -> IO ()
crawlCorpora opts = do
  cs <- getCorpora $ fetch opts
  case cs of
    Nothing -> do
      print "Failed parsing \"/corpora\"."
    Just cs' -> do
      mapM_ (crawlCorpus opts) $ catMaybes $ map ((fmap T.unpack) . crpsName) cs'

-- | Recursive crawling into a corpus
crawlCorpus :: Opts -> String -> IO ()
crawlCorpus opts corpusName = do
  case verbosity opts of
    Normal -> return ()
    Verbose -> do
      print $ "GET /corpora/" ++ corpusName
      return ()
  crps <- getCorpus (fetch opts) corpusName
  case crps of
    Nothing -> do
      print $ "Failed parsing \"/corpora/"  ++ corpusName ++ "\"."
    Just c -> do
      print $ "Crawling \"/corpora/"  ++ corpusName ++ "\"."
      let plays = map (T.unpack . mtdName . plyMetadata) $ crpsDramas c
      mapM_ (crawlPlay opts corpusName) plays
  threadDelay (delay opts)

crawlPlay :: Opts -> String -> String -> IO ()
crawlPlay opts corpusName playName = do
  case verbosity opts of
    Normal -> return ()
    Verbose -> do
      print $ "GET /corpora/" ++ corpusName ++ "/play/" ++ playName
      return ()
  ply <- getPlay (fetch opts) corpusName playName
  case ply of
    Nothing -> do
      print $ "Failed parsing \"corpora/" ++ corpusName ++ "/play/" ++ playName ++ "\"."
    Just _ -> return ()
  ply <- getPlayMetrics (fetch opts) corpusName playName
  case ply of
    Nothing -> do
      print $ "Failed parsing \"corpora/" ++ corpusName ++ "/play/" ++ playName ++ "/metrics\"."
    Just _ -> return ()
  threadDelay (delay opts)


src/Text/DraCor/API.hs
module Text.DraCor.API
  where

-- | Wrappers around a DraCor API

import qualified Data.ByteString.Lazy as B
import Data.Aeson

import Text.DraCor.Types
import Text.DraCor.TolerantJSON

-- * Public

getInfo :: (String -> IO B.ByteString) -> IO (Maybe Info)
getInfo f = do
  s <- f "/info"
  return $ (decode s :: Maybe Info)

getCorpora :: (String -> IO B.ByteString) -> IO (Maybe [Corpus])
getCorpora f = do
  s <- f "/corpora"
  return $ (decode s :: Maybe [Corpus])

getCorpus :: (String -> IO B.ByteString) -> String -> IO (Maybe Corpus)
getCorpus f corpus = do
  s <- f $ "/corpora/" ++ corpus
  return $ (decode s :: Maybe Corpus)

getCorpusMetadata :: (String -> IO B.ByteString) -> String -> IO (Maybe [Play])
getCorpusMetadata f corpus = do
  s <- f $ "/corpora/" ++ corpus ++ "/metadata"
  return $ (decode s :: Maybe [Play])

getPlay :: (String -> IO B.ByteString) -> String -> String -> IO (Maybe Play)
getPlay f corpus play = do
  s <- f $ "/corpora/" ++ corpus ++ "/play/" ++ play
  return $ (decode s :: Maybe Play)

getPlayMetrics :: (String -> IO B.ByteString) -> String -> String -> IO (Maybe Metrics)
getPlayMetrics f corpus play = do
  s <- f $ "/corpora/" ++ corpus ++ "/play/" ++ play ++ "/metrics"
  return $ (decode s :: Maybe Metrics)



src/Text/DraCor/Types.hs
{-# LANGUAGE DeriveGeneric #-}
module Text.DraCor.Types
  where

import Data.Text (Text)
import GHC.Generics


-- * Records for storing json data from the DraCor API



type URL = Text

type Year = Int

-- | 'Info' is a record for version information etc. about the API.
data Info = Info
  { apiName :: Text
  , apiStatus :: Text
  , apiExistdb :: Text
  , apiVersion :: Text
  } deriving (Generic, Show, Eq)


-- | A network node with node node-specific metrics.
data Node = Node
  { nodeId :: Text              -- ^ the ID of the node, must be given.
  , nodeWeightedDegree :: Maybe Int
  , nodeDegree :: Maybe Int
  , nodeCloseness :: Maybe Double
  , nodeEigenvector :: Maybe Double
  , nodeBetweenness :: Maybe Double
  } deriving (Show, Eq, Generic)


-- | A scene of a play and its speakers or persons on stage.
data Scene = Scene
  { scnNumber :: Int            -- ^ the running number of the scene, must be given
  , scnTitle :: Maybe Text
  , scnSpeakers :: [Text]
  , scnType :: Maybe Text
  } deriving (Generic, Show, Eq)


data Source
  = Source
  { srcName :: Maybe Text
  , srcUrl :: Maybe URL
  } 
  | SimpleSource { srcSource :: Text }
  deriving (Generic, Show, Eq)

sourceName :: Source -> Maybe Text
sourceName (Source n _) = n
sourceName (SimpleSource n) = Just n

sourceUrl :: Source -> Maybe URL
sourceUrl (Source _ u) = u
sourceUrl (SimpleSource _) = Nothing


data CastItem = CastItem
  { cstiId :: Maybe Text              -- ^ The ID used in the @who attribute of TEI
  , cstiName :: Maybe Text
  , cstiSex :: Maybe Text
  , cstiIsGroup :: Maybe Bool
  } deriving (Generic, Show, Eq)


data Author = Author
  { authrName :: Maybe Text
  , authrKey :: Maybe Text
  } deriving (Generic, Show, Eq)


data Metadata = Metadata
  { mtdId :: Text
  , mtdName :: Text
  , mtdCorpus :: Maybe Text
  -- , mtdPlayName :: Maybe Text -- is redundant
  , mtdGenre :: Maybe Text
  , mtdTitle :: Maybe Text
  , mtdSubtitle :: Maybe Text
  , mtdAuthors :: [Author]
  , mtdSource :: Maybe Text
  , mtdSourceUrl :: Maybe URL
  , mtdOriginalSource :: Maybe Text -- What's the difference to mdtSource?
  , mtdYearPremiered :: Maybe Year
  , mtdYearPrinted :: Maybe Year
  , mtdYearNormalized :: Maybe Year
  , mtdYearWritten :: Maybe Year
  , mtdWikidataId :: Maybe Text
  , mtdNetworkdataCsvUrl :: Maybe URL
  } deriving (Generic, Show, Eq)


-- | A record of a play's metrics.
data Metrics = Metrics
  { -- name, corpus and ID: Isn't it more haskell'isch to have it outside?
    mtrSize :: Maybe Int
  , mtrAverageClustering :: Maybe Double
  , mtrNumOfPersonGroups :: Maybe Int
  , mtrDensity :: Maybe Float
  , mtrAveragePathLength :: Maybe Float
  , mtrMaxDegreeIds :: Maybe [Text]  -- Maybe [Text]
  , mtrAverageDegree :: Maybe Float
  , mtrDiameter :: Maybe Int
  , mtrMaxDegree :: Maybe Int
  , mtrNumOfSpeakers :: Maybe Int
  , mtrNumConnectedComponents :: Maybe Int
  , mtrNumOfSpeakersFemale :: Maybe Int
  , mtrNumOfSpeakersMale :: Maybe Int
  , mtrNumOfSpeakersUnknown :: Maybe Int
  , mtrNumOfSegments :: Maybe Int
  , mtrNumOfActs :: Maybe Int
  , mtrNetworkSize :: Maybe Int
  , mtrAllInIndex :: Maybe Float
  , mtrAllInSegment :: Maybe Int
    -- Nodes
  , mtrNodes :: [Node]
  } deriving (Show, Eq, Generic)


data Play = Play
  { -- meta data
    plyMetadata :: Metadata
    -- Metrics
  , plyMetrics :: Metrics
    -- Scenes
  , plySegments :: [Scene]
    -- characters
  , plyCast :: [CastItem]
  } deriving (Generic, Show, Eq)


data Corpus = Corpus
  { crpsName :: Maybe Text
  , crpsTitle :: Maybe Text
  , crpsRepository :: Maybe Text
  , crpsUri :: Maybe Text
  , crpsDramas :: [Play]
  } deriving (Generic, Show, Eq)


src/Text/DraCor/CommonJSON.hs
{-# LANGUAGE TemplateHaskell, OverloadedStrings #-}
module Text.DraCor.CommonJSON where

-- | Common parsers an serializers used by the tolerant and intolerant
-- library.

import Data.Aeson
import Data.Maybe
import qualified Data.Text as T
import qualified Data.Aeson.TH as ATH
import Data.Foldable
import Data.Aeson.Types

import Text.DraCor.Types
import Text.DraCor.Utils


-- * JSON parser for 'Info'

$(ATH.deriveJSON ATH.defaultOptions{ATH.fieldLabelModifier = modifyField 3} ''Info)


-- * JSON for 'Scene'

--  $(ATH.deriveJSON ATH.defaultOptions{ATH.fieldLabelModifier = modifyField 3} ''Scene)

instance FromJSON Scene where
  parseJSON = withObject "Scene" $ \s -> Scene
    <$> s .: "number"
    <*> (fmap (fmap T.pack) $ s .:? "title")
    <*> (s .:? "speakers" .!= [])
    <*> (fmap (fmap T.pack) $ s .:? "type")

instance ToJSON Scene


-- * JSON for 'CastItem'

$(ATH.deriveJSON ATH.defaultOptions{ATH.fieldLabelModifier = modifyField 4} ''CastItem)


-- * JSON for 'Author'

$(ATH.deriveJSON ATH.defaultOptions{ATH.fieldLabelModifier = modifyField 5} ''Author)



src/Text/DraCor/IntolerantJSON.hs
{-# LANGUAGE TemplateHaskell #-}
module Text.DraCor.IntolerantJSON
  where

-- | This module defines a set of parsers and serializers for JSON
-- without workarounds for issues in the dracor API.

import Data.Aeson
import Data.Maybe
import qualified Data.Text as T
import qualified Data.Aeson.TH as ATH

import Text.DraCor.Types
import Text.DraCor.CommonJSON
import Text.DraCor.Utils


-- * JSON for 'Source'

$(ATH.deriveJSON ATH.defaultOptions{ATH.fieldLabelModifier = modifyField 3} ''Source)


-- * JSON for 'Node'

$(ATH.deriveJSON ATH.defaultOptions{ATH.fieldLabelModifier = modifyField 4} ''Node)


-- -- * JSON for 'Play'

-- $(ATH.deriveJSON ATH.defaultOptions{ATH.fieldLabelModifier = modifyField 3} ''Play)


-- -- * JSON for 'PlayFromCorpusList'

-- $(ATH.deriveJSON ATH.defaultOptions{ATH.fieldLabelModifier = modifyField 5} ''PlayFromCorpusList)


-- -- * JSON for 'Corpus'

-- $(ATH.deriveJSON ATH.defaultOptions{ATH.fieldLabelModifier = modifyField 4} ''Corpus)


src/Text/DraCor/TolerantJSON.hs
{-# LANGUAGE TemplateHaskell, OverloadedStrings #-}
module Text.DraCor.TolerantJSON
  where

-- | This module defines a fault tolerant set of parsers and
-- serializers for JSON. It has some work arounds for issues of the
-- dracor API.

import Data.Aeson
import Data.Aeson.Types (Parser)
import Data.Maybe
import qualified Data.Text as T
import qualified Data.Aeson.TH as ATH
import Data.Foldable
import Data.Aeson.Types
import Control.Monad
import Control.Applicative
import Data.Either
import Data.Text.Read

import Text.DraCor.Types
import Text.DraCor.CommonJSON
import Text.DraCor.Utils


-- * JSON for 'Source'

instance FromJSON Source where
  parseJSON = parseSource

parseSource :: Value -> Parser Source
parseSource (Object s) = Source
  <$> s .:? "name"
  <*> s .:? "url"
parseSource (String v) = SimpleSource
  <$> pure v
parseSource invalid =
  prependFailure "parsing \'source\' failed, " (typeMismatch "Object" invalid)

instance ToJSON Source


-- * JSON for 'Node'

instance ToJSON Node

instance FromJSON Node where
  parseJSON = parseNode

parseNode :: Value -> Parser Node
parseNode (Object o) = Node
  <$> o .: "id"
  -- workaround issue # of dracor-api
  <*> tolerateStringNum decimal o "weightedDegree" "weightedDegree" -- o .:? "weightedDegree"
  <*> tolerateStringNum decimal o "degree" "degree" -- o .:? "degree"
  <*> tolerateStringNum double o "closeness" "closeness" -- o .:? "closeness"
  <*> o .:? "eigenvector"
  -- workaroud issue # of dracor-api
  <*> tolerateStringNum double o "betweenness" "betweenness"


-- * JSON for 'Metrics'

instance FromJSON Metrics where
  parseJSON = parseMetrics

parseMetrics :: Value -> Parser Metrics
parseMetrics (Object s) = Metrics
  <$> s .:? "size"
  -- Workaround for issue #85 of dracor-api:
  -- <*> (fmap Just $ s .:? "averageClustering" .!= 0.0)
  <*> tolerateStringNum double s "averageClustering" "averageClustering"
  <*> s .:? "numOfPersonGroups"
  <*> s .:? "density"
  <*> s .:? "averagePathLength"
  -- Workaround for issue #84 of dracor-api:
  <*> (asum
        [ s .:? "maxDegreeIds" -- alread a list of strings
        , fmap (fmap (T.splitOn "|")) $ s .:? "maxDegreeIds"
        ])
  <*> s .:? "averageDegree"
  <*> s .:? "diameter"
  <*> s .:? "maxDegree"
  <*> s .:? "numOfSpeakers"
  <*> s .:? "numConnectedComponents"
  <*> s .:? "numOfSpeakersFemale"
  <*> s .:? "numOfSpeakersMale"
  <*> s .:? "numOfSpeakersUnknown"
  <*> s .:? "numOfSegments"
  <*> s .:? "numOfActs"
  -- workaround issue # of dracor-api
  <*> tolerateStringNum decimal s "networkSize" "networkSize" --  s .:? "networkSize"
  <*> s .:? "allInIndex"
  <*> s .:? "allInSegment"
  <*> s .:? "nodes" .!= []
parseMetrics _ = mzero

instance ToJSON Metrics


-- * JSON for 'Metadata'

instance FromJSON Metadata where
  parseJSON = parseMetadata

instance ToJSON Metadata

-- | Parser for 'Metadata'.
parseMetadata :: Value -> Parser Metadata
parseMetadata (Object s) = Metadata
  <$> s .: "id"
  <*> s .: "name"
  <*> s .:? "corpus"
  <*> s .:? "genre"
  <*> s .:? "title"
  <*> s .:? "subtitle"
  <*> s .:? "authors" .!= []
  -- Workaround for issue #88 of dracor-api:
  <*> (fmap (sourceName) $
       ((explicitParseFieldMaybe parseSource s "source") .!= (Source Nothing Nothing)))
  -- Workaround for issue #88 of dracor-api:
  <*> asum [ fmap Just $ s .: "sourceUrl" -- not .:?, because we want it failable
           , (fmap (sourceUrl) $
              ((explicitParseFieldMaybe parseSource s "source") .!= (Source Nothing Nothing)))
           ]
  <*> s .:? "originalSource"
  -- Workaround for issue #83 of dracor-api:
  <*> tolerateStringNum (signed decimal) s "yearPremiered" "premiereYear"
  <*> tolerateStringNum (signed decimal) s "yearPrinted" "printYear"
  <*> tolerateStringNum (signed decimal) s "yearNormalized" "normalizedYear"
  <*> tolerateStringNum (signed decimal) s "yearWritten" "writtenYear"
  <*> s .:? "wikidataId"
  <*> s .:? "networkdataCsvUrl"
parseMetadata _ = mzero
  
-- | Parse a numerical value or a string representing a numerical
-- value (in an other field).
tolerateStringNum :: (FromJSON a) =>
                     Reader a   -- ^ a reader from Data.Text.Read
                  -> Object     -- ^ the json object
                  -> T.Text     -- ^ the name of the numerical field
                  -> T.Text     -- ^ name of (alternative) text field
                  -> Parser (Maybe a)
tolerateStringNum reader v numField strField = asum
  [ fmap Just $ v .: numField -- not .:?, because we want it to fail if not present
  , fmap (join . (fmap ((either (pure Nothing) (Just . fst)) . reader))) $
    (v .:? strField :: Parser (Maybe T.Text))
  ]


-- * JSON for 'Play'

--  $(ATH.deriveJSON ATH.defaultOptions{ATH.fieldLabelModifier = modifyField 3} ''Play)


instance ToJSON Play

instance FromJSON Play where
  parseJSON = parsePlay
  
parsePlay :: Value -> Parser Play
parsePlay v@(Object o) = Play
  <$> parseMetadata v
  <*> parseMetrics v
  <*> o .:? "segments" .!= []
  <*> o .:? "cast" .!= []
parsePlay _ = mzero


-- * JSON for 'Corpus'

instance ToJSON Corpus

instance FromJSON Corpus where
  parseJSON = parseCorpus

parseCorpus :: Value -> Parser Corpus
parseCorpus (Object o) = Corpus
  <$> o .:? "name"
  <*> o .:? "title"
  <*> o .:? "repository"
  <*> o .:? "uri"
  <*> o .:? "dramas" .!= []
parseCorpus _ = mzero
